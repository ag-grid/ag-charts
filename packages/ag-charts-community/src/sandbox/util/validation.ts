import { Logger } from '../../util/logger';
import { isArray, isBoolean, isFiniteNumber, isFunction, isObject, isString } from '../../util/type-guards';
import { stringify } from '../../util/validation';
import { joinFormatted } from './string';

const descriptionSymbol = Symbol('description');
const requiredSymbol = Symbol('required');

type ObjectLikeDef<T> = T extends object ? (keyof T extends never ? never : OptionsDefs<T>) : never;

// Definitions for options validation with support for nested structures.
export type OptionsDefs<T> = { [K in keyof T]-?: Validator | ObjectLikeDef<T[K]> };

// Validator interface with optional description and required flag for better error messages.
export interface Validator extends Function {
    (value: unknown): boolean;
    [descriptionSymbol]?: string;
    [requiredSymbol]?: boolean;
}

/**
 * Validates the provided options object against the specified options definitions. Logs warnings for any invalid options encountered.
 * @param options The options object to validate.
 * @param optionsDefs The definitions against which to validate the options.
 * @param path (Optional) The current path in the options object, for nested properties.
 * @returns A boolean indicating whether the options are valid.
 */
export function validation<T>(options: T, optionsDefs: OptionsDefs<T>, path = '') {
    const extendPath = (key: string) => (isArray(optionsDefs) ? `${path}[${key}]` : `${path}.${key}`);
    for (const [key, validatorOrDefs] of Object.entries<Validator | ObjectLikeDef<any>>(optionsDefs)) {
        const value = options[key as keyof T];
        if (isFunction(validatorOrDefs)) {
            if (validatorOrDefs(value)) continue;

            let description = validatorOrDefs[descriptionSymbol];
            description = description ? `; expecting ${description}` : '';

            Logger.warn(
                `Option \`${extendPath(key)}\` cannot be set to [${stringify(value)}]${description}, ignoring.`
            );

            return false;
        } else if (!validation(value, validatorOrDefs, extendPath(key))) {
            return false;
        }
    }
    return true;
}

/**
 * Attaches a descriptive message to a validator function.
 * @param validator The validator function to which to attach a description.
 * @param description The description to attach.
 * @returns A new validator function with the attached description.
 */
export function attachDescription(validator: Validator, description: string): Validator {
    return Object.assign((value: unknown) => validator(value), { [descriptionSymbol]: description });
}

/**
 * Marks a validator or option definitions object as required.
 * @param validatorOrDefs The validator or option definitions to mark as required.
 * @returns The modified validator or option definitions, marked as required.
 */
export function required<T extends OptionsDefs<any>>(validatorOrDefs: T): T;
export function required<T extends OptionsDefs<any>[]>(validatorOrDefs: T): T;
export function required(validatorOrDefs: Validator): Validator;
export function required<T>(validatorOrDefs: T): T {
    return Object.assign(
        isFunction(validatorOrDefs)
            ? (value: any) => (validatorOrDefs as Function)(value)
            : optionsDefs(validatorOrDefs as OptionsDefs<any>),
        { [requiredSymbol]: true }
    ) as T;
}

/**
 * Creates a validator for ensuring an object matches the provided option definitions.
 * @param defs The option definitions against which to validate an object.
 * @param description (Optional) A description for the validator, defaulting to 'an object'.
 * @returns A validator function for the given option definitions.
 */
export const optionsDefs = <T>(defs: OptionsDefs<T>, description = 'an object'): Validator =>
    attachDescription(
        (value: unknown) =>
            isObject(value) &&
            Object.entries<Validator | ObjectLikeDef<any>>(defs).every(([key, validatorOrDefs]) =>
                isFunction(validatorOrDefs) ? validatorOrDefs(value[key]) : optionsDefs(validatorOrDefs)
            ),
        description
    );

/**
 * Combines multiple validators, requiring all to pass.
 * @param validators An array of validators to combine.
 * @returns A validator that requires all specified validators to pass.
 */
export const and = (...validators: Validator[]) =>
    attachDescription(
        (value: unknown) => validators.every((validator) => validator(value)),
        validators
            .map((v) => v[descriptionSymbol])
            .filter(Boolean)
            .join(' and ')
    );

/**
 * Combines multiple validators, passing if any one of them does.
 * @param validators An array of validators to combine.
 * @returns A validator that passes if any one of the specified validators does.
 */
export const or = (...validators: Validator[]) =>
    attachDescription(
        (value: unknown) => validators.some((validator) => validator(value)),
        validators
            .map((v) => v[descriptionSymbol])
            .filter(Boolean)
            .join(' or ')
    );

// Base type validators with descriptions.
export const array = attachDescription(isArray, 'an array');
export const boolean = attachDescription(isBoolean, 'a boolean');
export const callback = attachDescription(isFunction, 'a function');
export const number = attachDescription(isFiniteNumber, 'a number');
export const object = attachDescription(isObject, 'an object');
export const string = attachDescription(isString, 'a string');

// Numeric type validators with specific conditions.
export const positiveNumber = attachDescription((value) => isFiniteNumber(value) && value >= 0, 'a positive number');
export const ratio = attachDescription(
    (value) => isFiniteNumber(value) && value >= 0 && value <= 1,
    'a number between 0 and 1 inclusive'
);

/**
 * Creates a validator for a union of allowed values.
 * @param allowed An array of allowed values.
 * @returns A validator function that checks if a value is among the allowed ones.
 */
export const union = (...allowed: any[]) =>
    attachDescription(
        (value: any) => allowed.includes(value),
        `a keyword such as ${joinFormatted(allowed, 'or', (value) => `'${value}'`)}`
    );

/**
 * Creates a validator for a single constant value.
 * @param allowed The allowed constant value.
 * @returns A validator function that checks for equality with the allowed value.
 */
export const constant = (allowed: boolean | number | string) =>
    attachDescription((value: any) => allowed === value, `the value ${JSON.stringify(allowed)}`);

/**
 * Creates a validator for instances of a specific class.
 * @param instanceType The constructor of the class to check instances against.
 * @returns A validator function that checks if a value is an instance of the specified class.
 */
export const instanceOf = (instanceType: Function) =>
    attachDescription((value) => value instanceof instanceType, `an instance of ${instanceType.name}`);

/**
 * Creates a validator for arrays where every element must pass a given validator.
 * @param validator The validator to apply to each array element.
 * @returns A validator function for arrays with elements validated by the specified validator.
 */
export const arrayOf = (validator: Validator) =>
    attachDescription(
        (value: unknown) => isArray(value) && value.every(validator),
        `${validator[descriptionSymbol]} array`
    );
