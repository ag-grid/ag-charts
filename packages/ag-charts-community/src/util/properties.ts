import { extractDecoratedPropertyMetadata, listDecoratedProperties } from './decorator';
import { Logger } from './logger';
import { isArray } from './type-guards';

export class BaseProperties<T extends object = object> {
    set(properties: T) {
        const { className = this.constructor.name } = this.constructor as { className?: string };

        if (typeof properties !== 'object') {
            Logger.warn(`unable to set ${className} - expecting a properties object`);
            return this;
        }

        const keys = new Set(Object.keys(properties));
        for (const propertyKey of listDecoratedProperties(this)) {
            if (keys.has(propertyKey)) {
                const value = properties[propertyKey as keyof T];
                const self = this as any;
                if (isProperties(self[propertyKey])) {
                    // re-set property to force re-validation
                    self[propertyKey] =
                        self[propertyKey] instanceof PropertiesArray
                            ? self[propertyKey].reset(value)
                            : self[propertyKey].set(value);
                } else {
                    self[propertyKey] = value;
                }
                keys.delete(propertyKey);
            }
        }
        for (const unknownKey of keys) {
            Logger.warn(`unable to set [${unknownKey}] in ${className} - property is unknown`);
        }

        return this;
    }

    isValid<TContext = Omit<T, 'type'>>(this: TContext) {
        return listDecoratedProperties(this).every((propertyKey) => {
            const { optional } = extractDecoratedPropertyMetadata(this, propertyKey)!;
            // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
            const valid = optional || typeof this[propertyKey as keyof TContext] !== 'undefined';
            if (!valid) {
                Logger.warnOnce(`[${propertyKey}] is required.`);
            }
            return valid;
        });
    }

    toJson<J>(this: J): T {
        return listDecoratedProperties(this).reduce<Record<string, any>>((object, propertyKey) => {
            const propertyValue = this[propertyKey as keyof J];
            object[propertyKey] = isProperties(propertyValue) ? propertyValue.toJson() : propertyValue;
            return object;
        }, {}) as T;
    }
}

export class PropertiesArray<T extends BaseProperties> extends Array {
    private readonly itemFactory!: (params: object) => T;

    constructor(itemFactory: (new () => T) | ((params: object) => T), ...properties: object[]) {
        super(properties.length);
        const isConstructor = (value: Function): value is new () => T => Boolean(value?.prototype?.constructor?.name);
        const value = isConstructor(itemFactory) ? (params: object) => new itemFactory().set(params) : itemFactory;
        Object.defineProperty(this, 'itemFactory', { value, enumerable: false, configurable: false });
        this.set(properties);
    }

    set(properties: object[]): PropertiesArray<T> {
        if (isArray(properties)) {
            this.length = properties.length;
            for (let i = 0; i < properties.length; i++) {
                this[i] = this.itemFactory(properties[i]);
            }
        }
        return this;
    }

    reset(properties: object[]): PropertiesArray<T> {
        return new PropertiesArray(this.itemFactory, ...properties);
    }

    toJson() {
        return this.map((value) => value?.toJson?.() ?? value);
    }
}

export class TypedPropertiesArray<T extends BaseProperties> extends PropertiesArray<T> {
    private readonly itemFactories!: { [type: string]: new () => T };

    constructor(itemFactories: { [type: string]: new () => T }, ...properties: { type: string }[]) {
        super(itemFactories[Object.keys(itemFactories)[0]], ...properties);
        Object.defineProperty(this, 'itemFactories', { value: itemFactories, enumerable: false, configurable: false });
        this.set(properties);
    }

    override set(properties: { type: string }[]): TypedPropertiesArray<T> {
        if (!isArray(properties)) return this;

        this.length = properties.length;
        for (let i = 0; i < properties.length; i++) {
            const factory = this.itemFactories[properties[i].type];
            if (!factory) {
                Logger.warnOnce(
                    `Can not set property of unknown type [${properties[i].type}], expected one of [${Object.keys(this.itemFactories)}], ignoring.`
                );
                continue;
            }
            this[i] = new factory().set(properties[i]);
        }

        return this;
    }

    override reset(properties: { type: string }[]): TypedPropertiesArray<T> {
        return new TypedPropertiesArray(this.itemFactories, ...properties);
    }
}

export function isProperties<T extends object>(value: unknown): value is BaseProperties<T> {
    return value instanceof BaseProperties || value instanceof PropertiesArray;
}
