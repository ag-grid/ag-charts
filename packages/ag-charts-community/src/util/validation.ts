/* eslint-disable sonarjs/no-duplicate-string */
import { Color } from './color';
import { addTransformToInstanceProperty, BREAK_TRANSFORM_CHAIN } from './decorator';
import { Logger } from './logger';
type ValidationContext = { target: any };

export type ValidatePredicate = {
    (v: any, ctx: ValidationContext): boolean;
    message?: string;
};

export function Validate(predicate: ValidatePredicate) {
    return addTransformToInstanceProperty((target, prop, v: any) => {
        if (predicate(v, { target })) {
            return v;
        }

        const cleanKey = prop.toString().replace(/^_*/, '');

        let targetClass = target.constructor?.className ?? target.constructor?.name;
        if (targetClass?.length < 3) {
            targetClass = null;
        }

        const targetClassName = targetClass ? `of [${targetClass}] ` : '';
        if (predicate.message) {
            Logger.warn(
                `Property [${cleanKey}] ${targetClassName}cannot be set to [${JSON.stringify(v)}]; ${
                    predicate.message
                }, ignoring.`
            );
        } else {
            Logger.warn(`Property [${cleanKey}] ${targetClassName}cannot be set to [${JSON.stringify(v)}], ignoring.`);
        }

        return BREAK_TRANSFORM_CHAIN;
    });
}

export function predicateWithMessage(predicate: ValidatePredicate, message: string): ValidatePredicate {
    predicate.message = message;
    return predicate;
}

export const OPTIONAL = (v: any, ctx: ValidationContext, predicate: ValidatePredicate) =>
    v === undefined || predicate(v, ctx);

export const ARRAY = (length?: number, predicate?: ValidatePredicate) => {
    return predicateWithMessage(
        (v: any, ctx) =>
            Array.isArray(v) &&
            (length ? v.length === length : true) &&
            (predicate ? v.every((e) => predicate(e, ctx)) : true),
        `expecting an Array`
    );
};
export const OPT_ARRAY = (length?: number) => {
    return predicateWithMessage((v: any, ctx) => OPTIONAL(v, ctx, ARRAY(length)), 'expecting an optional Array');
};

export const AND = (...predicates: ValidatePredicate[]) => {
    return predicateWithMessage(
        (v: any, ctx) => predicates.every((p) => p(v, ctx)),
        predicates
            .map((p) => p.message)
            .filter((m) => m != null)
            .join(' AND ')
    );
};
export const OR = (...predicates: ValidatePredicate[]) => {
    return predicateWithMessage(
        (v: any, ctx) => predicates.some((p) => p(v, ctx)),
        predicates
            .map((p) => p.message)
            .filter((m) => m != null)
            .join(' OR ')
    );
};

const isComparable = (v: any) => {
    return v != null && !isNaN(v);
};
export const LESS_THAN = (otherField: string) =>
    predicateWithMessage(
        (v: number | Date, ctx) =>
            !isComparable(v) || !isComparable(ctx.target[otherField]) || v < ctx.target[otherField],
        `expected to be less than ${otherField}`
    );
export const GREATER_THAN = (otherField: string) =>
    predicateWithMessage(
        (v: number | Date, ctx) =>
            !isComparable(v) || !isComparable(ctx.target[otherField]) || v > ctx.target[otherField],
        `expected to be greater than ${otherField}`
    );

export const FUNCTION = predicateWithMessage((v: any) => typeof v === 'function', 'expecting a Function');
export const OPT_FUNCTION = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, FUNCTION),
    `expecting an optional Function`
);

export const BOOLEAN = predicateWithMessage((v: any) => v === true || v === false, 'expecting a Boolean');
export const OPT_BOOLEAN = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, BOOLEAN),
    'expecting an optional Boolean'
);

export const STRING = predicateWithMessage((v: any) => typeof v === 'string', 'expecting a String');
export const OPT_STRING = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, STRING),
    'expecting an optional String'
);

export const DATE = predicateWithMessage((v: any) => v instanceof Date && !isNaN(+v), 'expecting a Date object');
export const OPT_DATE = predicateWithMessage((v: any, ctx) => OPTIONAL(v, ctx, DATE), 'expecting an optional Date');
export const DATE_ARRAY = predicateWithMessage(ARRAY(undefined, DATE), 'expecting an Array of Date objects');

export const DATETIME_MS = NUMBER(0);
export const OPT_DATETIME_MS = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, DATETIME_MS),
    'expecting an optional number'
);

export const OPT_DATE_OR_DATETIME_MS = OR(OPT_DATE, OPT_DATETIME_MS);

const colorMessage = `A color string can be in one of the following formats to be valid: #rgb, #rrggbb, rgb(r, g, b), rgba(r, g, b, a) or a CSS color name such as 'white', 'orange', 'cyan', etc`;

export const COLOR_STRING = predicateWithMessage((v: any) => {
    if (typeof v !== 'string') {
        return false;
    }

    return Color.validColorString(v);
}, `expecting a color String. ${colorMessage}`);
export const OPT_COLOR_STRING = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, COLOR_STRING),
    `expecting an optional color String. ${colorMessage}`
);

export const COLOR_STRING_ARRAY = predicateWithMessage(
    ARRAY(undefined, COLOR_STRING),
    `expecting an Array of color strings. ${colorMessage}`
);
export const OPT_COLOR_STRING_ARRAY = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, COLOR_STRING_ARRAY),
    `expecting an optional Array of color strings. ${colorMessage}`
);

export function NUMBER(min?: number, max?: number) {
    const message = `expecting a finite Number${
        (min !== undefined ? ', more than or equal to ' + min : '') +
        (max !== undefined ? ', less than or equal to ' + max : '')
    }`;
    return predicateWithMessage(
        (v: any) =>
            typeof v === 'number' &&
            Number.isFinite(v) &&
            (min !== undefined ? v >= min : true) &&
            (max !== undefined ? v <= max : true),
        message
    );
}
export function OPT_NUMBER(min?: number, max?: number) {
    const message = `expecting an optional finite Number${
        (min !== undefined ? ', more than or equal to ' + min : '') +
        (max !== undefined ? ', less than or equal to ' + max : '')
    }`;
    return predicateWithMessage((v: any, ctx) => OPTIONAL(v, ctx, NUMBER(min, max)), message);
}

export function NUMBER_OR_NAN(min?: number, max?: number) {
    // Can be NaN or finite number
    const message = `expecting a finite Number${
        (min !== undefined ? ', more than or equal to ' + min : '') +
        (max !== undefined ? ', less than or equal to ' + max : '')
    }`;

    return predicateWithMessage(
        (v: any) =>
            typeof v === 'number' &&
            (isNaN(v) ||
                (Number.isFinite(v) && (min !== undefined ? v >= min : true) && (max !== undefined ? v <= max : true))),
        message
    );
}

export const NUMBER_ARRAY = predicateWithMessage(ARRAY(undefined, NUMBER()), 'expecting an Array of numbers');
export const OPT_NUMBER_ARRAY = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, NUMBER_ARRAY),
    'expecting an optional Array of numbers'
);

export const STRING_ARRAY = predicateWithMessage(ARRAY(undefined, STRING), 'expecting an Array of strings');
export const OPT_STRING_ARRAY = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, STRING_ARRAY),
    'expecting an optional Array of strings'
);
export function STRING_UNION(...values: string[]) {
    const message = `expecting one of: ${values.join(', ')}`;

    return predicateWithMessage((v: any) => typeof v === 'string' && values.indexOf(v) >= 0, message);
}

export const BOOLEAN_ARRAY = predicateWithMessage(ARRAY(undefined, BOOLEAN), 'expecting an Array of boolean values');
export const OPT_BOOLEAN_ARRAY = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, BOOLEAN_ARRAY),
    'expecting an optional Array of boolean values'
);

const FONT_WEIGHTS = [
    'normal',
    'bold',
    'bolder',
    'lighter',
    '100',
    '200',
    '300',
    '400',
    '500',
    '600',
    '700',
    '800',
    '900',
];

export const FONT_STYLE = predicateWithMessage(
    (v: any) => v === 'normal' || v === 'italic' || v === 'oblique',
    `expecting a font style keyword such as 'normal', 'italic' or 'oblique'`
);
export const OPT_FONT_STYLE = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, FONT_STYLE),
    `expecting an optional font style keyword such as 'normal', 'italic' or 'oblique'`
);

export const FONT_WEIGHT = predicateWithMessage(
    (v: any) => FONT_WEIGHTS.includes(v),
    `expecting a font weight keyword such as 'normal', 'bold' or 'bolder' or a numeric value such as 100, 300 or 600`
);
export const OPT_FONT_WEIGHT = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, FONT_WEIGHT),
    `expecting an optional font weight keyword such as 'normal', 'bold' or 'bolder' or a numeric value such as 100, 300 or 600`
);

export const LINE_DASH = predicateWithMessage(
    ARRAY(undefined, NUMBER(0)),
    'expecting an Array of numbers specifying the length in pixels of alternating dashes and gaps, for example, [6, 3] means dashes with a length of 6 pixels with gaps between of 3 pixels.'
);
export const OPT_LINE_DASH = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, LINE_DASH),
    'expecting an optional Array of numbers specifying the length in pixels of alternating dashes and gaps, for example, [6, 3] means dashes with a length of 6 pixels with gaps between of 3 pixels.'
);

const LINE_CAPS = ['butt', 'round', 'square'];
export const LINE_CAP = predicateWithMessage(
    (v: any) => LINE_CAPS.includes(v),
    `expecting a line cap keyword such as 'butt', 'round' or 'square'`
);
export const OPT_LINE_CAP = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, LINE_CAP),
    `expecting an optional line cap keyword such as 'butt', 'round' or 'square'`
);

const LINE_JOINS = ['round', 'bevel', 'miter'];
export const LINE_JOIN = predicateWithMessage(
    (v: any) => LINE_JOINS.includes(v),
    `expecting a line join keyword such as 'round', 'bevel' or 'miter'`
);
export const OPT_LINE_JOIN = predicateWithMessage(
    (v: any, ctx) => OPTIONAL(v, ctx, LINE_JOIN),
    `expecting an optional line join keyword such as 'round', 'bevel' or 'miter'`
);

const POSITIONS = ['top', 'right', 'bottom', 'left'];
export const POSITION = predicateWithMessage(
    (v: any) => POSITIONS.includes(v),
    `expecting a position keyword such as 'top', 'right', 'bottom' or 'left`
);

const INTERACTION_RANGES = ['exact', 'nearest'];
export const INTERACTION_RANGE = predicateWithMessage(
    (v: any) => (typeof v === 'number' && Number.isFinite(v)) || INTERACTION_RANGES.includes(v),
    `expecting an interaction range of 'exact', 'nearest' or a number`
);

const TEXT_WRAPS = ['never', 'always', 'hyphenate', 'on-space'];
export const TEXT_WRAP = predicateWithMessage(
    (v: any) => TEXT_WRAPS.includes(v),
    `expecting a text wrap strategy keyword such as 'never', 'always', 'hyphenate', 'on-space'`
);
