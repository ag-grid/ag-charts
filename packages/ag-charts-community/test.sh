#!/bin/bash

set -u

SCRIPT_DIR=$(cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd)
PROJECTS=""

if [[ $(basename $(pwd)) == "ag-charts-enterprise" ]] ; then
  FONTDIR=$PWD/src/test
else
  FONTDIR=$PWD/src/chart/test
fi

function testDev {
    npm run test:lint
    export FC_DEBUG=0
    export FONTCONFIG_PATH=$FONTDIR
    export FONTCONFIG_FILE=$FONTDIR/fonts.conf
    export PANGOCAIRO_BACKEND=fontconfig
    npx jest ${PROJECTS} --detectOpenHandles ${1:-}
}

function testCI {
    export FC_DEBUG=0
    export FONTCONFIG_PATH=$FONTDIR
    export FONTCONFIG_FILE=$FONTDIR/fonts.conf
    export PANGOCAIRO_BACKEND=fontconfig
    npx jest --runInBand
}

if [[ "$OSTYPE" != "linux-gnu"* ]] ; then
    echo -e "\033[41m\033[30mLinux is the only supported execution environment for Charts e2e tests.\033[0m"
    PROJECTS="--selectProjects unit"
fi

if [[ "${TEAMCITY_PROJECT_NAME:-}" == "" ]] ; then
  case ${1:-} in
    watch)
      testDev --watch
    ;;

    coverage)
      testDev --coverage
    ;;

    update)
      export SNAPSHOT_FAILURE_THRESHOLD=0.001
      testDev -u
    ;;

    *)
      testDev
    ;;
  esac
else
  testCI
fi
