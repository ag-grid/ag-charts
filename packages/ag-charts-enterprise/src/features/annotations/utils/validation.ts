import { type Direction, type _ModuleSupport, _Util } from 'ag-charts-community';

import type { AnnotationAxisContext, AnnotationContext, Point } from '../annotationTypes';

const { Logger } = _Util;

export function validateDatumLine(
    context: AnnotationContext,
    datum: { start: Point; end: Point },
    warningPrefix?: string
) {
    let valid = true;

    valid &&= validateDatumPoint(context, datum.start, warningPrefix && `${warningPrefix}[start] `);
    valid &&= validateDatumPoint(context, datum.end, warningPrefix && `${warningPrefix}[end] `);

    return valid;
}

export function validateDatumValue(
    context: AnnotationContext,
    datum: { value?: string | number | Date; direction?: Direction },
    warningPrefix: string
) {
    const axis = datum.direction === 'horizontal' ? context.yAxis : context.xAxis;
    const valid = validateDatumPointDirection(datum.value, axis);

    if (!valid && warningPrefix) {
        Logger.warnOnce(`${warningPrefix}is outside the axis domain, ignoring. - value: [${datum.value}]]`);
    }

    return valid;
}

export function validateDatumPoint(context: AnnotationContext, point: Point, warningPrefix?: string) {
    if (point.x == null || point.y == null) {
        if (warningPrefix) {
            Logger.warnOnce(`${warningPrefix}requires both an [x] and [y] property, ignoring.`);
        }
        return false;
    }

    const validX = validateDatumPointDirection(point.x, context.xAxis);
    const validY = validateDatumPointDirection(point.y, context.yAxis);

    if (!validX || !validY) {
        let text = 'x & y domains';
        if (validX) text = 'y domain';
        if (validY) text = 'x domain';
        if (warningPrefix) {
            Logger.warnOnce(`${warningPrefix}is outside the ${text}, ignoring. - x: [${point.x}], y: ${point.y}]`);
        }
        return false;
    }

    return true;
}

export function validateDatumPointDirection(value: any, context: AnnotationAxisContext) {
    const domain = context.scaleDomain();
    if (domain && context.continuous) {
        return value >= domain[0] && value <= domain.at(-1);
    }
    return true; // domain.includes(value); // TODO: does not work with dates
}

export function isPoint(point: Point | undefined): point is Point {
    return point != null;
}
