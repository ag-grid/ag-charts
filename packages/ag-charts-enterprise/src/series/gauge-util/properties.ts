import { _ModuleSupport } from 'ag-charts-community';

const { MARKER_SHAPE, UNION, OR } = _ModuleSupport;

export const FILL_MODE = UNION(['continuous', 'discrete'], 'a fill mode');
export const TARGET_MARKER_SHAPE = OR(MARKER_SHAPE, UNION(['line'], 'a marker shape'));
export const CORNER_MODE = UNION(['container', 'item'], 'a corner mode');
