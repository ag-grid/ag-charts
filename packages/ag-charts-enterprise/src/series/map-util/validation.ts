import { _ModuleSupport } from 'ag-charts-community';

function isValidCoordinate(v: any) {
    return Array.isArray(v) && v.length >= 2 && v.every(_ModuleSupport.isFiniteNumber);
}

function isValidCoordinates(v: any) {
    return Array.isArray(v) && v.length >= 2 && v.every(isValidCoordinate);
}

function hasSameStartEndPoint(c: number[][]) {
    return Math.abs(c[0][0] - c[c.length - 1][0]) < 1e-3 && Math.abs(c[0][1] - c[c.length - 1][1]) < 1e-3;
}

function isValidPolygon(v: any) {
    return Array.isArray(v) && v.every(isValidCoordinates) && v.every(hasSameStartEndPoint);
}

function isValidGeometry(v: any) {
    if (v === null || typeof v !== 'object' || v.type == null) return false;

    const { type, coordinates } = v;

    // Point, LineString, Polygon, MultiPoint, MultiLineString, and MultiPolygon
    switch (type) {
        case 'MultiPolygon':
            return Array.isArray(coordinates) && coordinates.every(isValidPolygon);
        case 'Polygon':
            return isValidPolygon(coordinates);
        case 'MultiLineString':
            return Array.isArray(coordinates) && coordinates.every(isValidCoordinates);
        case 'LineString':
            return isValidCoordinates(coordinates);
        case 'MultiPoint':
            return isValidCoordinates(coordinates);
        case 'Point':
            return isValidCoordinate(coordinates);
    }
}

function isValidFeature(v: any) {
    return v !== null && typeof v === 'object' && v.type === 'Feature' && isValidGeometry(v.geometry);
}

function isValidFeatureCollection(v: any) {
    return (
        v !== null &&
        typeof v === 'object' &&
        v.type === 'FeatureCollection' &&
        Array.isArray(v.features) &&
        v.features.every(isValidFeature)
    );
}

export const GEOJSON_OBJECT = _ModuleSupport.predicateWithMessage(isValidFeatureCollection, 'a GeoJSON object');
