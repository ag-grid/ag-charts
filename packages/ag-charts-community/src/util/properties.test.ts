import { BaseProperties, PropertiesArray, isProperties } from './properties';
import { expectWarning, setupMockConsole } from './test/mockConsole';
import { NUMBER, OBJECT_ARRAY, STRING, Validate } from './validation';

describe('BaseProperties', () => {
    setupMockConsole();

    it('should correctly set properties on an instance', () => {
        class MyClass extends BaseProperties<{ prop1: string; prop2: number }> {
            @Validate(STRING)
            prop1!: string;

            @Validate(NUMBER)
            prop2!: number;
        }
        const instance = new MyClass();
        instance.set({ prop1: 'value1', prop2: 42 });
        expect(instance.prop1).toBe('value1');
        expect(instance.prop2).toBe(42);
    });

    it('should warn on setting unknown properties', () => {
        class MyClass extends BaseProperties<{ prop1: string }> {
            @Validate(STRING)
            prop1!: string;
        }
        const instance = new MyClass();
        instance.set({ unknownProp: 'value' } as any);
        expectWarning('AG Charts - unable to set [unknownProp] in MyClass - property is unknown');
    });

    it('should warn on providing non-object properties', () => {
        class MyClass extends BaseProperties<{ prop1: string }> {
            @Validate(STRING)
            prop1!: string;
        }
        const instance = new MyClass();
        instance.set('string' as any);
        expectWarning('AG Charts - unable to set MyClass - expecting a properties object');
    });

    it('should validate required properties correctly', () => {
        class MyClass extends BaseProperties<{ prop1: string; prop2?: number }> {
            @Validate(STRING)
            prop1!: string;

            @Validate(NUMBER, { optional: true })
            prop2?: number;
        }
        const instance = new MyClass().set({ prop1: 'value1' });
        expect(instance.isValid()).toBe(true);

        const incompleteInstance = new MyClass();
        expect(incompleteInstance.isValid()).toBe(false);
    });

    it('should serialize to JSON correctly', () => {
        class MyClass extends BaseProperties<{ prop1: string; prop2: number }> {
            @Validate(STRING)
            prop1!: string;

            @Validate(NUMBER)
            prop2!: number;
        }
        const instance = new MyClass().set({ prop1: 'value1', prop2: 42 });
        const json = instance.toJson();
        expect(json).toEqual({ prop1: 'value1', prop2: 42 });
    });
});

describe('PropertiesArray', () => {
    it('should correctly handle arrays of BaseProperties instances', () => {
        class MyProperties extends BaseProperties<{ value: string }> {
            @Validate(STRING)
            value!: string;
        }
        const array = new PropertiesArray(MyProperties, { value: 'item1' }, { value: 'item2' });
        expect(array[0]).toBeInstanceOf(MyProperties);
        expect(array[1]).toBeInstanceOf(MyProperties);
        expect(array[0].value).toBe('item1');
        expect(array[1].value).toBe('item2');
        expect(array.toJson()).toEqual([{ value: 'item1' }, { value: 'item2' }]);
    });

    it('should correctly handle arrays properties on BaseProperties instances', () => {
        class MyProperties extends BaseProperties<{ value: string }> {
            @Validate(STRING)
            value!: string;
        }
        class MyClass extends BaseProperties<{ props: { value: string }[] }> {
            @Validate(OBJECT_ARRAY)
            props = new PropertiesArray(MyProperties);
        }
        const instance = new MyClass().set({ props: [{ value: 'item1' }, { value: 'item2' }] });
        expect(instance.props).toBeInstanceOf(PropertiesArray);
        expect(instance.props[0]).toBeInstanceOf(MyProperties);
        expect(instance.props[1]).toBeInstanceOf(MyProperties);
        expect(instance.props[0].value).toBe('item1');
        expect(instance.props[1].value).toBe('item2');
        expect(instance.toJson()).toEqual({ props: [{ value: 'item1' }, { value: 'item2' }] });
    });

    it('should reset correctly', () => {
        class MyProperties extends BaseProperties<{ value: string }> {
            @Validate(STRING)
            value!: string;
        }
        const array = new PropertiesArray(MyProperties, { value: 'item1' });
        const newArray = array.reset([{ value: 'newItem' }]);
        expect(newArray[0]).toBeInstanceOf(MyProperties);
        expect(newArray[0].value).toBe('newItem');
        expect(newArray.length).toBe(1);
    });
});

describe('isProperties', () => {
    it('should identify BaseProperties and PropertiesArray instances correctly', () => {
        class MyProperties extends BaseProperties<{ value: string }> {
            value!: string;
        }
        const instance = new MyProperties();
        const array = new PropertiesArray(MyProperties);
        expect(isProperties(instance)).toBe(true);
        expect(isProperties(array)).toBe(true);
        expect(isProperties({})).toBe(false);
    });
});
