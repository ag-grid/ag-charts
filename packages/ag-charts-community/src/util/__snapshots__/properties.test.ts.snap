// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseProperties should validate required properties correctly 1`] = `
[
  [
    "AG Charts - [prop1] is required.",
  ],
]
`;

exports[`BaseProperties should warn on providing non-object properties 1`] = `
[
  [
    "AG Charts - unable to set MyClass - expecting a properties object",
  ],
]
`;

exports[`BaseProperties should warn on setting unknown properties 1`] = `
[
  [
    "AG Charts - unable to set [unknownProp] in MyClass - property is unknown",
  ],
]
`;
